use-cases/eurac/trainer.py

```diff
@@ -0,0 +1,532 @@
+                    self.model, self.val_loader, device, opt=None
+                )
+
+            # gather losses from each worker and place them on the main worker.
+            worker_val_losses = self.strategy.gather(val_loss, dst_rank=0)
+
+            if not self.strategy.is_main_worker:
+                continue
+
+            # Moving them all to the cpu() before performing calculations
+            worker_val_losses = [wvl.cpu() for wvl in worker_val_losses]
+            avg_val_loss = torch.mean(torch.stack(worker_val_losses)).detach().cpu()
+            self.lr_scheduler.step(avg_val_loss)
+            loss_history["train"].append(train_loss)
+            loss_history["val"].append(avg_val_loss)
+            self.log(
+                item=train_loss.item(),
+                identifier="train_loss_per_epoch",
+                kind="metric",
+                step=epoch,
+            )
+            self.log(
+                item=avg_val_loss.item(),
+                identifier="val_loss_per_epoch",
+                kind="metric",
+                step=epoch,
+            )
+
+            for target in trainer.P.target_names:
+                metric_history[f"train_{target}"].append(train_metric[target])
+                metric_history[f"val_{target}"].append(val_metric[target])
+
+            # Aggregate and log metrics
+            avg_metrics = pd.DataFrame(metric_history).mean().to_dict()
+            for m_name, m_val in avg_metrics.items():
+                self.log(
+                    item=m_val,
+                    identifier=m_name + "_epoch",
+                    kind="metric",
+                    step=epoch,
+                )
+
+            if avg_val_loss < best_loss:
+                best_loss = avg_val_loss
+                print(f"train loss: {train_loss}")
+                print(f"val loss: {avg_val_loss}")
+                best_model = deepcopy(self.model.state_dict())
+
+            epoch_end_time = timer()
+            epoch_time_tracker.add_epoch_time(
+                epoch - 1, epoch_end_time - epoch_start_time
+            )
+
+        if self.strategy.is_main_worker:
+            print(best_model)
+        # logging the best model
+        # self.log(
+        #     item=best_model,
+        #     identifier='best_model',
+        #     kind='model'
+        # )
+
+            # Report training metrics of last epoch to Ray
+            train.report({"loss": avg_val_loss.item(),
+                          "train_loss": train_loss.item()})
+
+        return loss_history, metric_history
+
+    def create_dataloaders(self, train_dataset, validation_dataset, test_dataset):
+        sampling_kwargs = {}
+        if isinstance(self.strategy, HorovodStrategy):
+            sampling_kwargs["num_replicas"] = self.strategy.global_world_size()
+            sampling_kwargs["rank"] = self.strategy.global_rank()
+
+        if isinstance(self.strategy, NonDistributedStrategy):
+            processing = "single-gpu"
```

Comment:
The Hython code doesn't support any other options than "single-gpu" and "multi-gpu", so even though I agree that it is a little unclear what is going on, this is because of the Hython code and not this use case's code